Introduce and port over dependency injection articles for Microsoft Extensions (#20758)
dotnet/docs: docs/core/extensions/dependency-injection-guidelines.md

---
title: Dependency injection guidelines
description: Learn various dependency injection guidelines and best practices for .NET application development.
author: IEvangelist
ms.author: dapine
ms.date: 09/23/2020
ms.topic: guide
---

# Dependency injection guidelines

> In this article, you'll learn general guidelines and best practices for implementing dependency injection in your .NET applications. From design principles to proper disposal patterns, thread safety, and service lifetime management - these recommendations are tried and true.

Reviewer: ```suggestion
This article provides general guidelines and best practices for implementing dependency injection in .NET applications.
```

## Design services for dependency injection

When designing services for dependency injection:

- Avoid stateful, static classes and members. Avoid creating global state by designing apps to use singleton services instead.
- Avoid direct instantiation of dependent classes within services. Direct instantiation couples the code to a particular implementation.
- Make services small, well-factored, and easily tested.

> If a class has many injected dependencies, it might be a sign that the class has too many responsibilities and violates the [Single Responsibility Principle (SRP)](/dotnet/standard/modern-web-apps-azure-architecture/architectural-principles#single-responsibility). Attempt to refactor the class by moving some of its responsibilities into new classes. Keep in mind that Razor Pages page model classes and MVC controller classes should focus on UI concerns.

Reviewer: > Keep in mind that Razor Pages page model classes and MVC controller classes should focus on UI concerns.

Sounds like a comment for the ASP.NET Core edition of DI docs.

```suggestion
If a class has many injected dependencies, it might be a sign that the class has too many responsibilities and violates the [Single Responsibility Principle (SRP)](/dotnet/standard/modern-web-apps-azure-architecture/architectural-principles#single-responsibility). Attempt to refactor the class by moving some of its responsibilities into new classes.
```

:::code language="csharp" source="snippets/configuration/console-di-disposable/Program.cs" range="1-21,41-60":::

The debug console shows the following sample output after running:

```console
Scope 1...
ScopedDisposable.Dispose()
TransientDisposable.Dispose()

Scope 2...
ScopedDisposable.Dispose()
TransientDisposable.Dispose()

info: Microsoft.Hosting.Lifetime[0]
      Application started.Press Ctrl+C to shut down.
info: Microsoft.Hosting.Lifetime[0]
     Hosting environment: Production
info: Microsoft.Hosting.Lifetime[0]
     Content root path: .\configuration\console-di-disposable\bin\Debug\net5.0
info: Microsoft.Hosting.Lifetime[0]
     Application is shutting down...
SingletonDisposable.Dispose()
```

### Services not created by the service container

Consider the following code:

```csharp
public void ConfigureServices(IServiceCollection services)
{
    services.AddSingleton(new ExampleService());

    // ...
}
```

In the preceding code:

- The `ExampleService` instance is **not** created by the service container.
- The framework does **not** dispose of the services automatically.
- The developer is responsible for disposing the services.

> ### IDisposable guidance for Transient and shared instances

Reviewer: The text in these sections is hard to follow without sample code.
